File: MyJava/JRex/src/java/org/mozilla/jrex/io/JRexInputStream.java
/* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public
 * License Version 1.1 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of
 * the License at http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * Contributor(s):
 *   C.N Medappa <jrex_moz@yahoo.com><>
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the NPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the NPL, the GPL or the LGPL.
 *
 * ***** END LICENSE BLOCK ***** */

package org.mozilla.jrex.io;

import java.io.*;
import org.mozilla.jrex.exception.JRexException;
import org.mozilla.jrex.log.JRexL;

/**
* JRexInputStream is java wrapper arround native nsIInputStream.
* @author <a href="mailto:jrex_moz@yahoo.com">C.N.Medappa</a>
* @version 1.0
*/

public class JRexInputStream extends InputStream{
	private static final String className	= JRexInputStream.class.getName();

	static{
		try{
			initIDs();
		}catch(JRexException ex){
			JRexL.printStackTrace(ex);
			throw new RuntimeException(ex);
		}
	}
	/**
	* The varaible representing handle to the native nsIInputStream instance.
	*/
    private int streamPeer=0;

    /**
     * Creates a <code>JRexInputStream</code> given peer stream handle.
     * @param      streamPeer handle to the native nsIInputStream instance
     * @exception  JRexException if <code>streamPeer</code> is invalid.
     */
    public JRexInputStream(int streamPeer)throws JRexException {
		if(JRexL.on)JRexL.log(className,"**** JRexInputStream() -->streamPeer<"+streamPeer+">****");
		if(streamPeer<1)throw new JRexException("Invalid native stream handle.");
		this.streamPeer = streamPeer;
    }

	/**
	* Returns the number of bytes that can be read from this native inputstream without blocking.
	* @return     the number of bytes that can be read from this native inputstream without blocking.
	* @exception  IOException  if an I/O error or JRexException occurs.
	*/
    public int available() throws IOException{
		if(JRexL.on)JRexL.log(className,"**** available() -->I AM HERE****");
		try{
			return Available();
		}catch(JRexException ex){
			JRexL.printStackTrace(ex);
			throw new IOException("JRexException <"+ex.getMessage()+"> Occured");
		}
	}


    /**
     * Reads a byte of data from this input stream. This method blocks if no input is yet available.
     * @return     the next byte of data, or <code>-1</code> if the end of the stream is reached.
     * @exception  IOException  if an I/O error or JRexException occurs.
     */
    public int read() throws IOException {
		if(JRexL.on)JRexL.log(className,"**** read() -->I AM HERE****");
		try{
			return Read();
		}catch(JRexException ex){
			JRexL.printStackTrace(ex);
			throw new IOException("JRexException <"+ex.getMessage()+"> Occured");
		}
    }

    /**
     * Reads up to <code>buff.length</code> bytes of data from this input
     * stream into an array of bytes. This method blocks until some input
     * is available.
     *
     * @param      buff   the buffer into which the data is read.
     * @return     the total number of bytes read into the buffer, or
     *             <code>-1</code> if there is no more data because the end of
     *             the stream has been reached.
     * @exception  IOException  if an I/O error or JRexException occurs.
     */

    public int read(byte buff[], int off, int len) throws IOException {
		if(JRexL.on)JRexL.log(className,"**** read() --> buff<"+buff+"> off<"+off+"> len<"+len+"> ****");
		if (buff == null)
			throw new NullPointerException();
        else
        if ((off < 0) || (len < 0) || (off > buff.length)  ||((off + len) > buff.length))
            throw new IndexOutOfBoundsException();

        try{
			return Read(buff,off,len);
		}catch(JRexException ex){
			JRexL.printStackTrace(ex);
			throw new IOException("JRexException <"+ex.getMessage()+"> Occured");
		}
    }

	/**
	* Reads up to <code>b.length</code> bytes of data from this input
	* stream into an array of bytes. This method blocks until some input
	* is available.
	*
	* @param      buff   the buffer into which the data is read.
	* @return     the total number of bytes read into the buffer, or
	*             <code>-1</code> if there is no more data because the end of
	*             the stream has been reached.
	* @exception  IOException  if an I/O error or JRexException occurs.
	*/
    public int read(byte buff[]) throws IOException {
		if (buff == null)
			throw new NullPointerException();
		return read(buff, 0, buff.length);
    }

	/**
	* Mark is not supported, hence this function does nothing.
	*/
    public synchronized void mark(int readlimit) {
		if(JRexL.on)JRexL.log(className,"**** mark() --> readlimit<"+readlimit+"> Mark Not Supported!!! ****");
    }

	/**
	* Since mark is not supported, this throws IOException.
	*/
    public synchronized void reset() throws IOException {
		throw new IOException("Mark Not Supported!!!");
    }

	/**
	* skip is not supported, hence return <code>-1</code>.
	*/
    public long skip(long n) throws IOException {
		if(JRexL.on)JRexL.log(className,"**** skip() --> n<"+n+"> skip Not Supported!!! ****");
		return -1;
    }

	/**
	* Mark is not supported.
	*/
    public boolean markSupported() {
		return false;
    }

	/**
	* Closes this input stream and releases any system resources
	* associated with the stream.
	* @exception  IOException  if an I/O error or JRexException occurs.
	*/
	public void close() throws IOException {
		if(JRexL.on)JRexL.log(className,"**** close() --> streamPeer<"+streamPeer+">****");
		if(streamPeer>0)
		try{
			Close();
		}catch(JRexException ex){
			JRexL.printStackTrace(ex);
			throw new IOException("JRexException <"+ex.getMessage()+"> Occured");
		}finally{
			streamPeer=0;
		}
	}

	/**
	* Ensures that the <code>close</code> method of this input stream is
	* called when there are no more references to it.
	*
	* @see        #close()
	*/
    protected void finalize() throws IOException  {
		if(JRexL.on)JRexL.log(className,"**** finalize() --> streamPeer<"+streamPeer+">****");
		if(streamPeer>0)
			try{
				close();
			}catch(IOException ioex){
				JRexL.printStackTrace(ioex);
			}
    }

    private static native void initIDs()throws JRexException;
    private native int Available()throws JRexException;
    private native int Read() throws JRexException;
    private native int Read(byte b[], int off, int len) throws JRexException;
    private native void Close()throws JRexException;

}

File: MyJava/JRex/src/native/JRexInputStream.cpp
/* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
/* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public
 * License Version 1.1 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of
 * the License at http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * Contributor(s):
 *   C.N Medappa <jrex_moz@yahoo.com><>
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the NPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the NPL, the GPL or the LGPL.
 *
 * ***** END LICENSE BLOCK ***** */

#include "JRexInputStream.h"
using namespace JREX_JNI_UTIL;

NS_IMPL_THREADSAFE_ISUPPORTS1(JRexInputStream, nsIInputStream)

JRexInputStream::JRexInputStream(jobject stream):mIsClosed(PR_FALSE){
    JNIEnv *env = JRex_GetEnv(0);
	JREX_LOGLN("JRexInputStream()--> **** env<"<<env<<"> ****")
    if(env && stream)
		mJavaInputStream=env->NewGlobalRef(stream);
	else{
		JREX_LOGLN("JRexInputStream()--> **** CREATION FAILED!!!!!!!!!****")
		throw NS_ERROR_NULL_POINTER;
	}
	JREX_LOGLN("JRexInputStream()--> **** mJavaInputStream<"<<mJavaInputStream<<"> ****")
}

JRexInputStream::~JRexInputStream(){
	JNIEnv *env = JRex_GetEnv(0);
    JREX_LOGLN("~JRexInputStream()--> **** env<"<<env<<"> ****")
    if(env && mJavaInputStream){
		if(!mIsClosed)Close();
		JREX_LOGLN("~JRexInputStream()--> **** deleting global ref ****")
		env->DeleteGlobalRef(mJavaInputStream);
	}
	JREX_LOGLN("~JRexInputStream()--> **** Done ****")
}

NS_IMETHODIMP JRexInputStream::Close(){
	mIsClosed=PR_TRUE;
	JNIEnv *env = JRex_GetEnv(0);
	JREX_LOGLN("Close()--> **** env<"<<env<<"> ****")
	if(IS_NULL(env))return NS_ERROR_FAILURE;

	env->CallVoidMethod(mJavaInputStream,closeMID);
	JNI_IF_EXP_DESC_CLR_RET("Close()--> **** Failed with Exception ****",NS_ERROR_FAILURE)
	JREX_LOGLN("Close()--> DONE ****")
	return NS_OK;
}


NS_IMETHODIMP JRexInputStream::Available(PRUint32 *_retval){
	NS_ENSURE_ARG_POINTER(_retval);
	*_retval=0;
	JNIEnv *env = JRex_GetEnv(0);
	JREX_LOGLN("Available()--> **** env<"<<env<<"> ****")
	if(IS_NULL(env))return NS_ERROR_FAILURE;

	jint available=env->CallIntMethod(mJavaInputStream,availableMID);
	JREX_LOGLN("Available()--> **** available<"<<available<<"> ****")
	JNI_IF_EXP_DESC_CLR_RET("Available()--> **** Failed with Exception ****",NS_ERROR_FAILURE)
	*_retval=available;
    JREX_LOGLN("Available()--> **** *_retval<"<<*_retval<<"> ****")
    return NS_OK;
}


NS_IMETHODIMP JRexInputStream::Read(char *aBuf, PRUint32 aCount, PRUint32 *_retval){
	NS_ENSURE_ARG_POINTER(aBuf);
	NS_ENSURE_ARG_POINTER(_retval);

	*_retval=0;
	JNIEnv *env = JRex_GetEnv(0);
	JREX_LOGLN("Read()--> **** env<"<<env<<"> aCount<"<<aCount<<">****")
	if(IS_NULL(env))return NS_ERROR_FAILURE;

	jboolean isCopy=JNI_FALSE;
	jbyte* memArr=NULL;
	jbyteArray myArray=NULL;
	jint bytesRead=0;

	myArray=env->NewByteArray(aCount);
	bytesRead=env->CallIntMethod(mJavaInputStream,readMID,myArray,0,(jint)aCount);
	JREX_LOGLN("Read()--> **** bytesRead <"<<bytesRead<<"> ****")
	JNI_IF_EXP_DESC_CLR_GT("Read()--> **** CallIntMethod read Failed with Exception ****",clean)

	memArr=env->GetByteArrayElements(myArray,&isCopy);
    memcpy(aBuf,memArr,bytesRead);
	env->ReleaseByteArrayElements(myArray,memArr,JNI_ABORT);
	env->DeleteLocalRef(myArray);
    *_retval=bytesRead<0?0:bytesRead;
    return NS_OK;
   clean:
   JREX_LOGLN("Read()--> **** BRANCHED TO CLEAN ****")
	if(myArray)
		env->DeleteLocalRef(myArray);
   	*_retval=0;
   	return NS_OK;
}


NS_IMETHODIMP JRexInputStream::ReadSegments(nsWriteSegmentFun aWriter, void * aClosure, PRUint32 aCount, PRUint32 *_retval){
	NS_ENSURE_ARG_POINTER(_retval);

	JNIEnv *env = JRex_GetEnv(0);
	JREX_LOGLN("ReadSegments()--> **** env<"<<env<<"> aCount<"<<aCount<<">****")
	if(IS_NULL(env))return NS_ERROR_FAILURE;

	nsresult rv=NS_OK;
	PRUint32 rVal=0;

	PRUint32 avail =(PRUint32)env->CallIntMethod(mJavaInputStream, availableMID);
	JREX_LOGLN("ReadSegments()--> **** available count<"<<avail<<"> ****")
	JNI_IF_EXP_DESC_CLR_RET("ReadSegments()--> **** CallIntMethod available with Exception ****",NS_ERROR_FAILURE)
    if(avail<1){
		*_retval=0;
		return NS_OK;
	}
    PRUint32 count=aCount>avail?avail:aCount;
	JREX_LOGLN("ReadSegments()--> **** actaul count<"<<count<<"> ****")

	char* from = (char*)nsMemory::Alloc(count);
	if(!from){
		JREX_LOGLN("ReadSegments()--> **** Alloc FAILED!!! ****")
		rv= NS_ERROR_OUT_OF_MEMORY;
		goto clean;
	}

	rv=Read(from, count, &rVal);
	JREX_LOGLN("ReadSegments()--> **** rv<"<<rv<<"> rVal<"<<rVal<<"> ****")
	if(NS_FAILED(rv) || rVal<1)goto clean;

	rv = aWriter(this, aClosure, from, 0, rVal, _retval);
	JREX_LOGLN("ReadSegments()--> **** rv<"<<rv<<"> _retval<"<<*_retval<<">****")
	if(NS_FAILED(rv))// errors returned from the writer is not propogated.
		*_retval=0;
	nsMemory::Free(from);
	return NS_OK;
clean:
	JREX_LOGLN("ReadSegments()--> **** BRANCHED TO CLEAN ****")
	if(from)nsMemory::Free(from);
	*_retval=0;
	return rv;
}

NS_IMETHODIMP JRexInputStream::IsNonBlocking(PRBool *_retval){
	*_retval = PR_FALSE;
    return NS_OK;
}
